File: Homework/themoviedb/src/main/java/com/example/themoviedb/repository/RepositoryProvider.java
package com.example.themoviedb.repository;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.themoviedb.network.ApiFactory;
import com.example.themoviedb.network.clients.MovieDBClient;

public class RepositoryProvider {

    private static volatile RepositoryProvider sInstance;

    @Nullable
    private MovieRepository repository;

    private RepositoryProvider() {
    }

    public static RepositoryProvider get() {
        if (sInstance == null) {
            synchronized (RepositoryProvider.class) {
                if (sInstance == null) {
                    sInstance = new RepositoryProvider();
                }
            }
        }
        return sInstance;
    }

    @NonNull
    private static <T> T getServiceInstance(Class<T> clazz) {
        return ApiFactory.getRetrofitInstance().create(clazz);
    }

    @NonNull
    public MovieRepository provideRepository() {
        if (repository == null) {
            repository = new MovieRepositoryImpl(getServiceInstance(MovieDBClient.class));
        }
        return repository;
    }
}

File: Homework/themoviedb/src/main/java/com/example/themoviedb/repository/RepositoryProvider.kt
package com.example.themoviedb.repository

import com.example.themoviedb.network.ApiFactory
import com.example.themoviedb.network.clients.MovieDBClient

class RepositoryProvider private constructor() {

    private var repository: MovieRepository? = null

    fun provideRepository(): MovieRepository {
        if (repository == null) {
            repository = MovieRepositoryImpl(getServiceInstance(MovieDBClient::class.java))
        }
        return repository as MovieRepository
    }

    companion object {

        @Volatile
        private var sInstance: RepositoryProvider? = null

        fun get(): RepositoryProvider? {
            if (sInstance == null) {
                synchronized(RepositoryProvider::class.java) {
                    if (sInstance == null) {
                        sInstance = RepositoryProvider()
                    }
                }
            }
            return sInstance
        }

        private fun <T> getServiceInstance(clazz: Class<T>): T {
            return ApiFactory.retrofitInstance!!.create(clazz)
        }
    }
}